### PR TITLE
FEAT : Adding VPTQ quantization method to HFQuantizer

### PR DESCRIPTION
**Pull request by wejoncy**
# What does this PR do?

This PR is intended to add support for Extreme Low-bit Vector Post-Training Quantization(VPTQ) to the transformers library.

VPTQ is a novel Post-Training Quantization method that leverages Vector Quantization to high accuracy on LLMs at an extremely low bit-width (<2-bit). VPTQ can compress 70B, even the 405B model, to 1-2 bits without retraining and maintain high accuracy.. More details here: https://github.com/microsoft/vptq

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@SunMarc @younesbelkada @ArthurZucker

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Niiiiice! 🚀

**Comment by MekkCyber** [comment]
Awesome PR @wejoncy, Thank you for adding this amazing quantization method ! The integration looks very smooth🔥! I Just some minor comments

File: docs/source/en/quantization/vptq.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+rendered properly in your Markdown viewer.
+
+-->
+
+# VPTQ 
+
+> [!TIP]
+> Try VPTQ on [Hugging Face](https://huggingface.co/spaces/microsoft/VPTQ)!
+> Try VPTQ on [Google Colab](https://colab.research.google.com/github/microsoft/VPTQ/blob/main/notebooks/vptq_example.ipynb)!
+
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply docs/source/en/quantization/vptq.md:23]: Sure.

File: src/transformers/integrations/vptq.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    if not is_vptq_available():
+        raise ValueError("VPTQ is not available. Please install it with `pip install vptq`")
+
+    if not is_accelerate_available():
+        raise ValueError(
+            f"VPTQ requires Accelerate to be installed: `pip install 'accelerate>={ACCELERATE_MIN_VERSION}'`"
+        )
+
+    from accelerate import init_empty_weights
+    from vptq import VQuantLinear
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply src/transformers/integrations/vptq.py]: Good suggestion.
Thanks

File: src/transformers/integrations/vptq.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+if is_torch_available():
+    import torch.nn as nn
+
+
+def replace_with_vptq_linear(
+    model,
+    quantization_config=None,
+    current_key_name=None,
+    has_been_replaced=False,
+):
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply src/transformers/integrations/vptq.py:28]: Well, we used `modules_to_convert` in config json, it has all the layers which should to be quantized and include corresponding parameters.
But I add `modules_to_not_convert` in the function signature to allow user setting which layer should be excluded.

File: src/transformers/quantizers/quantizer_vptq.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+from packaging import version
+
+from .base import HfQuantizer
+
+
+if TYPE_CHECKING:
+    from ..modeling_utils import PreTrainedModel
+
+from ..integrations import replace_with_vptq_linear
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply src/transformers/quantizers/quantizer_vptq.py]: Sure. Done

File: src/transformers/quantizers/quantizer_vptq.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    def validate_environment(self, *args, **kwargs):
+        if not is_accelerate_available():
+            raise ImportError("Using `vptq` quantization requires Accelerate: `pip install accelerate`")
+
+        if not is_vptq_available():
+            raise ImportError("Using `vptq` quantization requires VPTQ: `pip install vptq`")
+        import vptq
+
+        assert version.Version(vptq.__version__) >= version.Version("0.0.4"), "vptq is avaiable since 0.0.4"
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply src/transformers/quantizers/quantizer_vptq.py]: Sounds good. Thanks.

File: src/transformers/utils/quantization_config.py [hunk 6]
```diff
@@ -994,6 +995,95 @@ def post_init(self):
...
+    """
+    This is a wrapper class about `vptq` parameters.
+
+    Args:
+        enable_proxy_error (`bool`, *optional*, defaults to `False`): calculate proxy error for each layer
+        config_for_layers (`Dict`, *optional*, defaults to `{}`): quantization params for each layer
+        kwargs (`Dict[str, Any]`, *optional*):
+            Additional parameters from which to initialize the configuration object.
+    """
+
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply src/transformers/utils/quantization_config.py:1064]: Yes, It's exactly true. Config file contains all specific parameters and it a bit huge.
We used this to customize different quantization parameters for different linear layers to have the best performance.
But we will try to figure out a way to allow a simplified configuration.

File: tests/quantization/vptq_integration/test_vptq.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+@slow
+@require_torch_gpu
+@require_vptq
+@require_accelerate
+class VptqTest(unittest.TestCase):
+    model_name = "VPTQ-community/Meta-Llama-3.1-8B-Instruct-v12-k65536-4096-woft"
+
+    input_text = "Hello my name is"
+    max_new_tokens = 32
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply tests/quantization/vptq_integration/test_vptq.py:58]: Sure.

File: tests/quantization/vptq_integration/test_vptq.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+    EXPECTED_OUTPUT = "Hello my name is Sarah and I am a 25 year old woman from the United States. I am a college graduate and I am currently working as a marketing specialist for a small"
+
+    device_map = "cuda"
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply tests/quantization/vptq_integration/test_vptq.py:62]: I have fixed the `CPU` device option availability check.  We will optimize the CPU kernel in the next VPTQ release.
After that I will open another PR to support CPU and add a `CPU` unit test.

File: src/transformers/quantizers/quantizer_vptq.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+    def __init__(self, quantization_config: QuantizationConfigMixin, **kwargs):
+        super().__init__(quantization_config, **kwargs)
+        self.quantization_config = quantization_config
+
+    def validate_environment(self, *args, **kwargs):
+        if not is_accelerate_available():
+            raise ImportError("Using `vptq` quantization requires Accelerate: `pip install accelerate`")
+
+        if not is_vptq_available():
```
Comment (MekkCyber) [inline src/transformers/quantizers/quantizer_vptq.py:49]: Just a small nit, now that `is_vptq_available` checks the version too, you can change the message to specify that `vptq` should be installed and it requires the min version you want

File: tests/quantization/vptq_integration/test_vptq.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+        with init_empty_weights():
+            model = AutoModelForCausalLM(config)
+        quantization_config = VptqConfig(modules_to_not_convert=modules_to_not_convert)
+        model = replace_with_vptq_linear(model, quantization_config=quantization_config)
+        nb_vptq_linear = 0
+        for module in model.modules():
+            if isinstance(module, VQuantLinear):
+                nb_vptq_linear += 1
+
+        self.assertEqual(nb_linears - 25, nb_vptq_linear)
```
Comment (MekkCyber) [inline tests/quantization/vptq_integration/test_vptq.py:189]: Just for a bit of clarity you can specify here in a comment that 25 corresponds to 24 `decoder.layers.{layer_idx}.fc1` + `lm_head` in `modules_to_not_convert`

**Comment by MekkCyber** [comment]
Thanks for iterating @wejoncy LGTM 🔥! Left only two small nits

File: src/transformers/utils/quantization_config.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MekkCyber) [reply src/transformers/utils/quantization_config.py:1064]: Can we, for each `Linear` layer like `q_proj` or `down_proj`, have a list of parameter values, and each parameter corresponds to a layer ?

File: src/transformers/quantizers/quantizer_vptq.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+        if torch_dtype is None:
+            if torch.cuda.is_available():
+                torch_dtype = torch.float16
+                logger.info(
+                    "CUDA available. Assuming VPTQ inference on GPU and loading the model in `torch.float16`. To overwrite it, set `torch_dtype` manually."
+                )
+            else:
+                torch_dtype = torch.float32
+                logger.info(
+                    "CUDA is unavailable. Assuming VPTQ inference on CPU and loading the model in `torch.float32`. To overwrite it, set `torch_dtype` manually."
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply src/transformers/quantizers/quantizer_vptq.py]: We are working CPU kernels, but let's mark it unavailable for CPU at first.

File: src/transformers/quantizers/quantizer_vptq.py (hunk 9 above)
Comment (wejoncy) [reply src/transformers/quantizers/quantizer_vptq.py:49]: Yes. we have VPTQ_MIN_VERSION=0.0.4 as default.

Good suggestion. I have added the minimal version in the error logs for guiding user clearly.

File: tests/quantization/vptq_integration/test_vptq.py (hunk 10 above)
Comment (wejoncy) [reply tests/quantization/vptq_integration/test_vptq.py:189]: Sounds good. Thanks

**Comment by MekkCyber** [comment]
Thanks for the updates @wejoncy ! Can't wait for the PR to be merged !

File: src/transformers/utils/quantization_config.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (wejoncy) [reply src/transformers/utils/quantization_config.py:1064]: Hi @MekkCyber  sorry, not quite following your suggestion, could you please explain a little bit more? 
What we did is for each `linear` layer, we give it a paraments dict, such as
```
model.decoder_model.layers.0.self_attn.q_proj: {"v_len":[-1, 8], .....},
model.decoder_model.layers.0.self_attn.k_proj: {"v_len":[-1, 8], .....},
.
.
.
model.decoder_model.layers.n.self_attn.q_proj: {"v_len":[-1, 8], .....},
```

**Review by SunMarc — APPROVED** [review]
Thanks the PR ! Excited to see more VTPQ models on the hub. Just a few nits. Thanks also for the nice tests and the documentation

**Comment by YangWang92** [comment]
> Thanks the PR ! Excited to see more VTPQ models on the hub. Just a few nits. Thanks also for the nice tests and the documentation

Thank @SunMarc, for your review! We are actively developing new models and algorithms, and we expect to release support for multimodal models (LLaMA 3.2 and Qwen2-VL) next month.

File: src/transformers/utils/quantization_config.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MekkCyber) [reply src/transformers/utils/quantization_config.py:1064]: Hi @wejoncy, sorry I wasn't clear, I meant smth like this for example for the `up_proj` linears : 
```
model.layers.mlp.up_proj: {
        "bias": [null, null, ..., null],
        "enable_norm": [true, true, false, ..., true],
        ...
        "num_centroids": [
          (-1, 65536), (-1, 65536), ...,
        ],
        ...
      },
```
So that each property contains a list of values, and each value corresponds to a layer, so the length of the lists, is the number of layers, and for the i-th layer, we access the `bias`property for example using : `model.layers.mlp.up_proj["bias"][i]`

File: docs/source/en/quantization/vptq.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+from transformers import AutoTokenizer, AutoModelForCausalLM
+
+quantized_model = AutoModelForCausalLM.from_pretrained(
+    "VPTQ-community/Meta-Llama-3.1-70B-Instruct-v16-k65536-65536-woft",
+    torch_dtype="auto", 
+    device_map="auto"
+)
+tokenizer = AutoTokenizer.from_pretrained("VPTQ-community/Meta-Llama-3.1-70B-Instruct-v16-k65536-65536-woft")
+```
+
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply docs/source/en/quantization/vptq.md:45]: Sure. Thanks for the advice.

File: src/transformers/utils/quantization_config.py [hunk 13]
```diff
@@ -994,6 +995,95 @@ def post_init(self):
...
+        enable_perm: to perm input_channel or not
+        group_num: how many single groups for vector-quantization
+        group_size:depends on out-features
+        indices_as_float: for Finetuning
+        is_indice_packed: should always be True
+        num_centroids: centriod numbers of clusters
+        num_res_centroids: ditto for residual
+        outlier_size: outliers
+        vector_lens: centroid vector length in quantization
+    """
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply src/transformers/utils/quantization_config.py:1013]: Sure.

File: docs/source/en/quantization/vptq.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+# VPTQ 
+
+> [!TIP]
+> Try VPTQ on [Hugging Face](https://huggingface.co/spaces/microsoft/VPTQ)!
+> Try VPTQ on [Google Colab](https://colab.research.google.com/github/microsoft/VPTQ/blob/main/notebooks/vptq_example.ipynb)!
+> Know more about VPTQ on [ArXiv](https://arxiv.org/pdf/2409.17066)!
+
+Vector Post-Training Quantization ([VPTQ](https://github.com/microsoft/VPTQ)) is a novel Post-Training Quantization method that leverages Vector Quantization to high accuracy on LLMs at an extremely low bit-width (<2-bit). VPTQ can compress 70B, even the 405B model, to 1-2 bits without retraining and maintain high accuracy.
+
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply docs/source/en/quantization/vptq.md:25]: Sure.

**Comment by wejoncy** [comment]
Hi @ArthurZucker ,
A friendly Ping. Could you please take a look this PR?  I will be actively changing the code if any suggestions. 
Thanks.

**Comment by YangWang92** [comment]
Hi @ArthurZucker,
Just a polite ping to check in—may I ask if there’s anything else we need to modify? Thank you!

Best regards,
Yang

**Comment by code30x58** [comment]
https://youtu.be/tKCQcM5vfac?feature=shared

**Review by ArthurZucker — APPROVED** [review]
Let's go! 🤗

**Comment by ArthurZucker** [comment]
Super sorry for the late review, I am a bit under water 😓

**Comment by YangWang92** [comment]
> Super sorry for the late review, I am a bit under water 😓

Thanks so much! And thanks to everyone for their help! Special thanks to @xianbaoqian for the support—super grateful! 😊

**Comment by MekkCyber** [comment]
Hey @YangWang92, i was going through the VPTQ paper, and I wanted to make sure that I understand correctly, so VPTQ is the same as GPTQ except that the quantization step is not linear but uses VQ instead, is that correct ?

**Comment by YangWang92** [comment]
> Hey @YangWang92, i was going through the VPTQ paper, and I wanted to make sure that I understand correctly, so VPTQ is the same as GPTQ except that the quantization step is not linear but uses VQ instead, is that correct ?

Hi @MekkCyber, 
Sorry for the late reply! I'm quite busy recent days. VPTQ is indeed quite similar to GPTQ in terms of modeling the optimization problem, as both use second-order optimization methods for quantization. The key difference is that VPTQ specifically addresses the additional research questions introduced by Vector Quantization.
